design/src/rfcs/rfc0016_body_callback_apis_v2.md

```diff
@@ -0,0 +1,389 @@
+            Inner::Sha1(ref mut callback) => callback.update(bytes)?,
+            Inner::Sha256(ref mut callback) => callback.update(bytes)?,
+        };
+
+        Ok(())
+    }
+
+    pub fn trailers(&self) -> Result<Option<HeaderMap<HeaderValue>>, BoxError> {
+        match &self.0 {
+            Inner::Sha256(callback) => callback.trailers(),
+            Inner::Crc32c(callback) => callback.trailers(),
+            Inner::Crc32(callback) => callback.trailers(),
+            Inner::Sha1(callback) => callback.trailers(),
+        }
+    }
+
+    // TODO I think I still need to account for the size of the trailer names.
+    //      Should I be doing that here?
+    pub fn size_hint(&self) -> SizeHint {
+        let checksum_size_in_bytes = match &self.0 {
+            Inner::Crc32(_) => Crc32Callback::size(),
+            Inner::Crc32c(_) => Crc32cCallback::size(),
+            Inner::Sha1(_) => Sha1Callback::size(),
+            Inner::Sha256(_) => Sha256Callback::size(),
+        };
+        let base64_encoded_checksum_size_in_bytes =
+            base64::encoded_length(checksum_size_in_bytes as usize);
+
+        SizeHint::with_exact(base64_encoded_checksum_size_in_bytes as u64)
+    }
+}
+```
+
+```rust
+// In aws-smithy-checksums/src/body.rs
+use crate::ChecksumCallback;
+
+use aws_smithy_http::body::SdkBody;
+use aws_smithy_http::header::append_merge_header_maps;
+
+use bytes::{Buf, Bytes};
+use http::{HeaderMap, HeaderValue};
+use http_body::{Body, SizeHint};
+use pin_project::pin_project;
+
+use std::error::Error as StdError;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+#[pin_project]
+pub struct ChecksumBody<InnerBody> {
+    #[pin]
+    inner: InnerBody,
+    #[pin]
+    checksum_callback: ChecksumCallback,
+}
+
+impl ChecksumBody<SdkBody> {
+    pub fn new(checksum_algorithm: &str, body: SdkBody) -> Self {
+        Self {
+            checksum_callback: ChecksumCallback::new(checksum_algorithm),
+            inner: body,
+        }
+    }
+
+    // TODO a more intelligent person could prob impl this with generics
+    fn poll_inner(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Bytes, Error>>> {
+        let this = self.project();
+        let inner = this.inner;
+        let mut checksum_callback = this.checksum_callback;
+
+        match inner.poll_data(cx) {
+            Poll::Ready(Some(Ok(mut data))) => {
+                let len = data.chunk().len();
+                let bytes = data.copy_to_bytes(len);
+
+                if let Err(e) = checksum_callback.update(&bytes) {
+                    return Poll::Ready(Some(Err(e)));
+                }
+
+                Poll::Ready(Some(Ok(bytes)))
+            }
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Ready(Some(Err(e))) => Poll::Ready(Some(Err(e))),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+pub type Error = Box<dyn StdError + Send + Sync>;
```

Comment:
SdkBody's `Error` type is public, so you could potentially reuse it directly: https://github.com/awslabs/smithy-rs/blob/e1e9a29d5db1330eb51c1e84e5c6c3acb9b4f65e/rust-runtime/aws-smithy-http/src/body.rs#L19